src/tbb/assert_impl.h

```diff
@@ -37,45 +38,59 @@ namespace tbb {
 namespace detail {
 namespace r1 {
 #endif
-// TODO: consider extension for formatted error description string
-static void assertion_failure_impl(const char* location, int line, const char* expression, const char* comment) {
-
-    std::fprintf(stderr, "Assertion %s failed (located in the %s function, line in file: %d)\n",
-        expression, location, line);
-
-    if (comment) {
-        std::fprintf(stderr, "Detailed description: %s\n", comment);
-    }
-#if _MSC_VER && _DEBUG
-    if (1 == _CrtDbgReport(_CRT_ASSERT, location, line, "tbb_debug.dll", "%s\r\n%s", expression, comment?comment:"")) {
-        _CrtDbgBreak();
-    } else
-#endif
-    {
-        std::fflush(stderr);
-        std::abort();
-    }
-}
 
-// Do not move the definition into the assertion_failure function because it will require "magic statics".
-// It will bring a dependency on C++ runtime on some platforms while assert_impl.h is reused in tbbmalloc 
-// that should not depend on C++ runtime
+// Do not move the definition into the assertion_failure_impl function because it will require "magic statics".
+// It will bring a dependency on C++ runtime on some platforms while assert_impl.h is reused in tbbmalloc
+// that should not depend on C++ runtime. For the same reason, we cannot use std::call_once here.
 static std::atomic<tbb::detail::do_once_state> assertion_state;
 
-void __TBB_EXPORTED_FUNC assertion_failure(const char* location, int line, const char* expression, const char* comment) {
+// TODO: consider extension for formatted error description string
+/* [[noreturn]] */ static void assertion_failure_impl(const char* location, int line, const char* expression, const char* comment) {
 #if __TBB_MSVC_UNREACHABLE_CODE_IGNORED
     // Workaround for erroneous "unreachable code" during assertion throwing using call_once
     #pragma warning (push)
     #pragma warning (disable: 4702)
 #endif
-    // We cannot use std::call_once because it brings a dependency on C++ runtime on some platforms 
-    // while assert_impl.h is reused in tbbmalloc that should not depend on C++ runtime
-    atomic_do_once([&](){ assertion_failure_impl(location, line, expression, comment); }, assertion_state);
+    atomic_do_once([&](){
+        std::fprintf(stderr, "Assertion %s failed (located in the %s function, line in file: %d)\n",
+            expression, location, line);
+
+        if (comment) {
+            std::fprintf(stderr, "Detailed description: %s\n", comment);
+        }
+#if _MSC_VER && _DEBUG
+        if (1 == _CrtDbgReport(_CRT_ASSERT, location, line, "tbb_debug.dll", "%s\r\n%s", expression, comment?comment:"")) {
```

Comment:
I can't understand why this code works only for tbb_debug.dll. Why can't it be called from tbbmalloc and tbbbind as well?

Reply:
Agreed offline not to change anything here, as this is unrelated to this PR.